### PR TITLE
Shooter intake

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelConstants.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkMaxLimitSwitch;
+import edu.greenblitz.robotName.Robot;
+import edu.greenblitz.robotName.RobotConstants;
+import edu.greenblitz.robotName.utils.motors.GBSparkMax;
+
+
+public class FunnelConstants {
+
+	public static final double ROLL_POWER = 1;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelConstants.java:12]: add EJECT_POWER instrad of saying -ROLL_POWER in Funnel class

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelFactory.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package edu.greenblitz.robotName.subsystems.shooter.Funnel;
+
+import edu.greenblitz.robotName.RobotConstants;
+
+public class FunnelFactory {
+	public static IFunnel create() {
+		return switch (RobotConstants.ROBOT_TYPE) {
+			case ROBOT_NAME -> new NeoFunnel();
+			case REPLAY -> new ReplayFunnel();
+			default -> new SimulationFunnel();
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelFactory.java:10]: add a simulation case

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+public class NeoFunnel implements IFunnel{
+	
+	private GBSparkMax motor;
+
+	private Debouncer debouncer;
+
+	public NeoFunnel() {
+		motor = new GBSparkMax(FunnelConstants.SparkMaxConfigs.FUNNEL_ID, CANSparkMaxLowLevel.MotorType.kBrushless);
+		motor.config(FUNNEL_CONFIG_OBJECT);
+		motor.getReverseLimitSwitch(SWITCH_TYPE).enableLimitSwitch(true);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel.java:19]: magic boolean??

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	
+	@Override
+	public void setPower(double power) {
+		setVoltage(power * RobotConstants.SimulationConstants.BATTERY_VOLTAGE);
+	}
+	
+	@Override
+	public void setVoltage(double voltage) {
+		appliedOutput = MathUtil.clamp(voltage, -RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE, RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel.java:37]: just add a MIN_MOTOR_VOLTAGE

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]
a bit natzi

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelFactory.java (hunk 2 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelFactory.java:10]: there is default

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel.java (hunk 4 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel.java:37]: why add lines for fun  ):> look its a guy with a hat

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelConstants.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelConstants.java:12]: i dont see why. its just adding lines for fun

Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/FunnelConstants.java:12]: > i dont see why. its just adding lines for fun

you will probably want to eject at a higher speed than rolling

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import edu.greenblitz.robotName.subsystems.shooter.Funnel.FunnelInputsAutoLogged;
+import edu.greenblitz.robotName.subsystems.shooter.Funnel.IFunnel;
+import edu.greenblitz.robotName.utils.motors.GBSparkMax;
+import edu.wpi.first.math.filter.Debouncer;
+
+import static edu.greenblitz.robotName.subsystems.shooter.Funnel.NeoFunnel.NeoFunnelConstants.*;
+
+public class NeoFunnel implements IFunnel {
+	
+	private GBSparkMax motor;
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java:13]: Why is it a NEO? do we need a lot of power here? A brushed motor will have less problems, work better and give the same result.
Wait until @NoamRosenberg08 Talks to mechanics but probably we will switch so Brushed

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+	private Debouncer debouncer;
+
+	public NeoFunnel() {
+		motor = new GBSparkMax(FUNNEL_ID, CANSparkMaxLowLevel.MotorType.kBrushless);
+		motor.config(FUNNEL_CONFIG_OBJECT);
+		motor.getReverseLimitSwitch(SWITCH_TYPE).enableLimitSwitch(IS_SENSOR_EXIST);
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java:20]: Why is it a limit switch? We still want the ability to take the object backwards to bring in to the endeffector so i dont think it should be a limit.

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnelConstants.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import edu.greenblitz.robotName.utils.motors.GBSparkMax;
+
+public class NeoFunnelConstants {
+    public static final int FUNNEL_ID = 0;
+
+    public static final boolean IS_SENSOR_EXIST = true;
+
+    public static final SparkMaxLimitSwitch.Type SWITCH_TYPE = SparkMaxLimitSwitch.Type.kNormallyClosed;
+
+    public static double DEBOUNCE_TIME_FOR_LIMIT_SWITCH = 0.5;
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnelConstants.java:15]: Thats half a second. I think that the others are at 0.2 or less. Check in Charrrged-Up

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel/SimulationFunnel.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+import static edu.greenblitz.robotName.subsystems.shooter.Funnel.SimulationFunnel.SimulationFunnelConstants.*;
+
+public class SimulationFunnel implements IFunnel {
+	
+	private FlywheelSim motorSimulation;
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel/SimulationFunnel.java:17]: Why not DCMotorSim?

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel/SimulationFunnel.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	public void setVoltage(double voltage) {
+		appliedOutput = MathUtil.clamp(voltage, -RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE, RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE);
+		motorSimulation.setInputVoltage(appliedOutput);
+	}
+	
+	@Override
+	public void updateInputs(FunnelInputsAutoLogged funnelInputs) {
+		funnelInputs.appliedOutput = appliedOutput;
+		funnelInputs.outputCurrent = motorSimulation.getCurrentDrawAmps();
+		funnelInputs.velocity = motorSimulation.getAngularVelocityRPM();
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/SimulationFunnel/SimulationFunnel.java:50]: Why do i care about the velocity of the funnel? It is just forward and backward at power nothing more.

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java (hunk 5 above)
Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/shooter/Funnel/NeoFunnel/NeoFunnel.java:14]: when we know what will be the motor ill change

**Review by bossyshrimp68 — APPROVED** [review]

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by aaa-png — APPROVED** [review]
UNICODE:4E00

**Review by itamaroryan — APPROVED** [review]

**Pull request merged**